### DISCUSSION TITLE
When I try to use a resolver, I receive the error 'Second definition of mutation(...) (Types::MutationType) is invalid, already configured with Types::MutationType.'

### DISCUSSION THREAD

**Discussion by kaku0225** [rmosolgo/graphql-ruby#4690] [Q&A] [answered]
I've been recently trying to use GraphQL and wanted to refactor the methods inside the query_type. However, I encountered an error, and despite trying various methods, I couldn't successfully extract them. I also haven't duplicated the use of mutations. Why is this error occurring?

Here is my code.

```
# app/graphql/resolvers/base.rb

module Resolvers
  class Base < GraphQL::Schema::Resolver
    argument_class Arguments::Base
  end
end
```

```
# app/graphql/resolvers/recommended_items.rb

module Resolvers
  class RecommendedItems < Resolvers::Base
    type String, null: false

    def resolve
      'recommended_items'
    end
  end
end
```

```
# app/graphql/types/query_type.rb

module Types
  class QueryType < Types::BaseObject
    include GraphQL::Types::Relay::HasNodeField
    include GraphQL::Types::Relay::HasNodesField

    field :recommended_items, resolver: Resolvers::RecommendedItems, description: "Items this user might like"
  end
end

```

```
# app/graphql/test_for_graphql_schema.rb

class TestForGraphqlSchema < GraphQL::Schema
  mutation(Types::MutationType)
  query(Types::QueryType)

  # For batch-loading (see https://graphql-ruby.org/dataloader/overview.html)
  use GraphQL::Dataloader

  # GraphQL-Ruby calls this when something goes wrong while running a query:
  def self.type_error(err, context)
    # if err.is_a?(GraphQL::InvalidNullError)
    #   # report to your bug tracker here
    #   return nil
    # end
    super
  end

  # Union and Interface Resolution
  def self.resolve_type(abstract_type, obj, ctx)
    # TODO: Implement this method
    # to return the correct GraphQL object type for `obj`
    raise(GraphQL::RequiredImplementationMissingError)
  end

  # Stop validating when it encounters this many errors:
  validate_max_errors(100)

  # Relay-style Object Identification:

  # Return a string UUID for `object`
  def self.id_from_object(object, type_definition, query_ctx)
    # For example, use Rails' GlobalID library (https://github.com/rails/globalid):
    object.to_gid_param
  end

  # Given a string UUID, find the object
  def self.object_from_id(global_id, query_ctx)
    # For example, use Rails' GlobalID library (https://github.com/rails/globalid):
    GlobalID.find(global_id)
  end
end
```

I apologize once again, as I couldn't make further attempts due to the lack of clarity about the root cause of the issue. I'm sorry for taking up your time.

**Comment by rmosolgo** [comment]
Hi! Sorry for the trouble. Yes, that's very strange. My first guess is that, somehow, `test_for_graphql_schema.rb` is being executed twice. If you add a debugging message to that file and try again, do you see the message printed twice? For example: 


```ruby
# app/graphql/test_for_graphql_schema.rb

class TestForGraphqlSchema < GraphQL::Schema
  puts "Loading #{self}"
  mutation(Types::MutationType)
```

**Comment by kaku0225** [comment]
Hi rmosolgo
Thank you for your response.

When I added `puts "Loading #{self}"` in `app/graphql/test_for_graphql_schema.rb` it seems that it didn't execute twice.

```ruby
# app/graphql/test_for_graphql_schema.rb

class TestForGraphqlSchema < GraphQL::Schema
  puts "Loading #{self}"
  p '1'
  mutation(Types::MutationType)
  p '2'
  query(Types::QueryType)
  p '3'
  ...
  ...
```
![截圖 2023-11-01 上午9 26 29](https://github.com/rmosolgo/graphql-ruby/assets/73519567/8c90b410-c2f8-4ff2-9526-cc22a4812883)

Additionally, there are times when other error messages suddenly appear, and I'm not entirely sure if the root cause might actually be this.
![截圖 2023-11-01 上午9 31 00](https://github.com/rmosolgo/graphql-ruby/assets/73519567/e58988ac-524a-4905-ac97-0aea9766b9a4)

**Reply by rmosolgo** [reply, marked as answer]
Thanks for sharing -- I wouldn't be surprised if some of those other error messages are really the underlying cause. For example, Rails might start loading the schema `.rb` file, but then encounter an error part-way through and stop loading it. In that case, it would be partially loaded: the schema class would be defined, but since Rails didn't _finish_ loading it, then Rails would try again, causing this error. 

I suggest stopping the development server, then starting it again (to make sure that nothing is loaded yet), then replicating this error. After that, scroll back to the _start_ of the server logs and find the first error message. Hopefully addressing _that_ error will make things work better. 

The error in your screenshot is that Rails can't find `Arguments::Base`. That's referenced in the new resolver class, but Rails can't seem to find that constant. Where is it defined in your application? Maybe you could change the reference or file structure so that Rails can load it and find it as expected.

**Reply by kaku0225** [reply]
Thank you very much for helping me resolve the issue. I followed your instructions and successfully solved the problem. I appreciate your assistance and apologize for taking your time.
